dali/test/python/test_detection_pipeline.py

```diff
@@ -13,147 +13,342 @@
 # limitations under the License.
 
 from __future__ import print_function
-from nvidia.dali.pipeline import Pipeline
-import nvidia.dali.ops as ops
-import nvidia.dali.types as types
+
 import argparse
-import numpy as np
+import itertools
+import os
 import random
+from math import sqrt, ceil
+
+import numpy as np
+import nvidia.dali.ops as ops
+import nvidia.dali.types as types
+from nvidia.dali.backend_impl import TensorListGPU
+from nvidia.dali.pipeline import Pipeline
+
+
+def coco_anchors():
+    anchors = []
+
+    fig_size = 300
+    feat_sizes = [38, 19, 10, 5, 3, 1]
+    feat_count = len(feat_sizes)
+    steps = [8., 16., 32., 64., 100., 300.]
+    scales = [21., 45., 99., 153., 207., 261., 315.]
+    aspect_ratios = [[2], [2, 3], [2, 3], [2, 3], [2], [2]]
+
+    fks = []
+    for step in steps:
+        fks.append(fig_size / step)
+
+    anchor_idx = 0
+    for idx in range(feat_count):
+        sk1 = scales[idx] / fig_size
+        sk2 = scales[idx + 1] / fig_size
+        sk3 = sqrt(sk1 * sk2)
+
+        all_sizes = [[sk1, sk1], [sk3, sk3]]
+
+        for alpha in aspect_ratios[idx]:
+            w = sk1 * sqrt(alpha)
+            h = sk1 / sqrt(alpha)
+            all_sizes.append([w, h])
+            all_sizes.append([h, w])
+
+        for sizes in all_sizes:
+            w = sizes[0]
+            h = sizes[1]
+
+            for i in range(feat_sizes[idx]):
+                for j in range(feat_sizes[idx]):
+                    cx = (j + 0.5) / fks[idx]
+                    cy = (i + 0.5) / fks[idx]
+
+                    cx = max(min(cx, 1.), 0.)
+                    cy = max(min(cy, 1.), 0.)
+                    w = max(min(w, 1.), 0.)
+                    h = max(min(h, 1.), 0.)
+
+                    anchors.append(cx - 0.5 * w)
+                    anchors.append(cy - 0.5 * h)
+                    anchors.append(cx + 0.5 * w)
+                    anchors.append(cy + 0.5 * h)
+
+                    anchor_idx = anchor_idx + 1
+    return anchors
+
+
+class DetectionPipeline(Pipeline):
+    def __init__(self, args, device_id, file_root, annotations_file):
+        super(DetectionPipeline, self).__init__(
+            args.batch_size, args.num_workers, device_id, args.prefetch, args.seed)
+
+        # Reading COCO dataset
+        self.input = ops.COCOReader(
+            file_root=file_root,
+            annotations_file=annotations_file,
+            shard_id=device_id,
+            num_shards=args.num_gpus,
+            ratio=True,
+            ltrb=True,
+            random_shuffle=True)
+
+        self.decode_cpu = ops.HostDecoder(device="cpu", output_type=types.RGB)
```

Comment:
do you need two decoders?

Reply:
The idea of this test is to check different ways of doing the same thing against each other.

Reply:
Theoretically we can go with CPU one and then transfer to the GPU for the GPU ops. But I agree with   @awolant.

Reply:
I wanted to test corresponding ops on the same exact data, so this pipeline is rather not optimized.